File: npm/src/commands/req.js
'use strict'
const path = require('path')
const os = require('os')
const { Command } = require('commander')
const { listREQs, moveREQ } = require('../generators/req')
const { t } = require('../i18n')
const { homedir } = require('../homedir');

const cmd = new Command('req')
cmd.description(t('req.description'))

cmd.command('new <title>')
  .description(t('req.new.description'))
  .action(async (title) => {
    const { input, select } = require('@inquirer/prompts')
    const generators = require('../generators/req')
    const adrGenerators = require('../generators/adr')

    const content = { title, motivation: '', criteria: '', dependsOnADRs: [] }

    if (process.stdin.isTTY) {
      // Form 1 — título + motivação
      content.title = await input({ message: t('req.new.prompt.title'), default: title })
      content.motivation = await input({ message: t('req.new.prompt.motivation'), default: '' })

      // Detectar domínios com base em título + motivação
      const probes = generators.detectDomains(content.title + ' ' + content.motivation)

      // Form 2 — critérios de aceite
      content.criteria = await input({ message: t('req.new.prompt.criteria'), default: '- [ ]\n- [ ]' })

      // Escopo dos ADR drafts desta sessão (uma única pergunta, vale para todos)
      const adrScope = await select({
        message: t('req.new.prompt.adrScope'),
        choices: [
          { name: t('req.new.prompt.adrScopeLocal'), value: 'local' },
          { name: t('req.new.prompt.adrScopeGlobal'), value: 'global' },
        ],
        default: 'local',
      })
      const adrDir = adrScope === 'global'
        ? path.join(homedir(), '.trackfw', 'adr')
        : require('../config').load().adrDirs[0]

      // Perguntas dinâmicas por probe
      const generatedADRs = []
      for (const probe of probes) {
        for (const question of probe.questions) {
          const choices = question.options.map(opt => ({
            name: opt.label,
            value: opt.adrSlug || '',
          }))
          const answer = await select({
            message: question.text,
            choices,
          })
          if (answer) {
            try {
              const basename = await adrGenerators.newADRDraft(answer, adrDir)
              if (basename) generatedADRs.push(basename)
            } catch (e) {
              console.warn(t('req.new.adrWarning', { slug: answer, message: e.message }))
            }
          }
        }
      }

      content.dependsOnADRs = [...new Set(generatedADRs)]
    }

    await generators.newREQ(content)

    if (content.dependsOnADRs.length > 0) {
      console.log(`\n${t('req.new.adrDraftsCreated')}`)
      content.dependsOnADRs.forEach(adr => console.log(`  -> ${adr}`))
      console.log(`\n${t('req.new.resolveADRs')}`)
    }
  })

cmd.command('list')
  .description(t('req.list.description'))
  .action(async () => {
    listREQs(require('../config').load())
  })

cmd.command('move <name> <status>')
  .description('Update a REQ status in place')
  .action(async (name, status) => {
    try {
      moveREQ(name, status)
    } catch (err) {
      console.error(`Error: ${err.message}`)
      process.exitCode = 1
    }
  })

module.exports = cmd

File: internal/commands/req.go
package commands

import (
	"fmt"
	"os"

	"github.com/charmbracelet/huh"
	cbterm "github.com/charmbracelet/x/term"
	"github.com/kgsaran/trackfw/internal/config"
	"github.com/kgsaran/trackfw/internal/generators"
	"github.com/spf13/cobra"
)

func newReqCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "req",
		Short: "Manage Requirements",
	}
	cmd.AddCommand(newReqNewCmd())
	cmd.AddCommand(newReqListCmd())
	return cmd
}

func newReqNewCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "new <title>",
		Short: "Create a new REQ",
		Args:  cobra.ExactArgs(1),
		RunE:  runReqNew,
	}
}

func runReqNew(_ *cobra.Command, args []string) error {
	content := generators.REQContent{Title: args[0]}

	// Detectar se stdin é TTY — wizard interativo somente em TTY
	if !cbterm.IsTerminal(uintptr(os.Stdin.Fd())) {
		return generators.NewREQ(content)
	}

	// Form 1 — coleta título + motivação
	form1 := huh.NewForm(
		huh.NewGroup(
			huh.NewInput().
				Title("Project requirement").
				Description("Describe what you want to build or change").
				Value(&content.Title),
			huh.NewInput().
				Title("Motivation").
				Description("Why is this requirement needed?").
				Value(&content.Motivation),
		),
	)
	if err := form1.Run(); err != nil {
		return fmt.Errorf("wizard: %w", err)
	}

	// Detectar probes com base em título + motivação
	intention := content.Title + " " + content.Motivation
	detectedProbes := generators.DetectDomains(intention)

	// Construir grupos do Form 2 — sem campos manuais de ADR/roadmap:
	// os vínculos com ADRs são estabelecidos automaticamente pelo probe discovery abaixo.
	groups := []*huh.Group{
		huh.NewGroup(
			huh.NewInput().
				Title("Acceptance Criteria").
				Description("List acceptance criteria, one per line").
				Value(&content.Criteria),
		),
	}

	// Slice de respostas — indexada para evitar bug de closure
	answers := make([]string, 0)
	type questionRef struct {
		options []generators.ProbeOption
	}
	var questionRefs []questionRef

	for _, probe := range detectedProbes {
		for _, question := range probe.Questions {
			answers = append(answers, "")
			questionRefs = append(questionRefs, questionRef{options: question.Options})
			idx := len(answers) - 1
			opts := make([]huh.Option[string], len(question.Options))
			for i, opt := range question.Options {
				opts[i] = huh.NewOption(opt.Label, opt.ADRSlug)
			}
			groups = append(groups, huh.NewGroup(
				huh.NewSelect[string]().
					Title(question.Text).
					Options(opts...).
					Value(&answers[idx]),
			))
		}
	}

	// Form 2 — critérios, links e probes
	form2 := huh.NewForm(groups...)
	if err := form2.Run(); err != nil {
		return fmt.Errorf("wizard: %w", err)
	}

	// Processar respostas das probes → gerar ADR Drafts
	var generatedADRs []string
	for i, answer := range answers {
		_ = questionRefs[i] // referência mantida para rastreabilidade futura
		if answer != "" {   // ADRSlug não-vazio = decisão pendente
			basename, err := generators.NewADRDraft(answer)
			if err != nil {
				fmt.Printf("warning: could not create ADR draft for %s: %v\n", answer, err)
				continue
			}
			generatedADRs = append(generatedADRs, basename)
		}
	}
	content.DependsOnADRs = uniqueStrings(generatedADRs)

	if err := generators.NewREQ(content); err != nil {
		return err
	}

	if len(content.DependsOnADRs) > 0 {
		fmt.Println("\nADR drafts created:")
		for _, adr := range content.DependsOnADRs {
			fmt.Printf("  -> %s\n", adr)
		}
		fmt.Println("\nResolve these ADRs (set Status: Accepted) before creating a roadmap.")
	}

	return nil
}

// uniqueStrings remove duplicatas mantendo a ordem de primeira ocorrência.
func uniqueStrings(ss []string) []string {
	seen := map[string]bool{}
	var result []string
	for _, s := range ss {
		if !seen[s] {
			seen[s] = true
			result = append(result, s)
		}
	}
	return result
}

func newReqListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list",
		Short: "List all REQs in docs/req/",
		RunE: func(cmd *cobra.Command, args []string) error {
			return generators.ListREQs(config.Load().REQDir)
		},
	}
}
